graph.cpp

```diff
@@ -0,0 +1,98 @@
+#include <iostream>
+#include <list>
+#include <map>
+using namespace std;
+
+
+class Node{
+    public:
+        int nodeID;
+        // Placeholder for 3D location info as I'm not sure how Godot handles it
+        tuple <double, double, double> location;
+        // Placeholder for formula data
+        int data;
+        list<Node*> neighbors;
+
+        Node(){
+            nodeID = 0;
+            location = make_tuple(0,0,0);
+            data = 0; 
+        }
+        Node(int nodeID, tuple<double, double, double> location, int data){
+            this->nodeID = nodeID;
+            this->location = location;
+            this->data = data;
+        }
+};
+
+class ProofGraph{
+    int nodeCount;
+    map<int, Node> nodeMap;
+    public:
+
+        // Constructor, creates empty proof graph
+        ProofGraph(){
+            nodeCount = 0;
+        }
+
+        void displayNodeList(){
+            for (auto& tempNode : nodeMap){
+                cout << tempNode.second.nodeID << ": ";
+                for(auto& i : tempNode.second.neighbors){
+                    cout << i->nodeID << " ";
+                }
+                cout << "\n";
+            }
+        }
+
+        void addNode(tuple<double, double, double> location, int data){
+            Node temp = Node(this->nodeCount, location, data);
+            nodeMap[nodeCount] = temp;
+            nodeCount++;
+            //cout << "Node Count " << nodeCount << "\n";
+        }
+
+        // Needs checks for node validity
+        void addEdge(int start, int end){
+            nodeMap[start].neighbors.push_back(&(nodeMap[end]));
+            nodeMap[end].neighbors.push_back(&(nodeMap[start]));
+        }
+
+        // Needs checks for node validity
+        void removeNode(int deleteID){
+            Node* temp = &nodeMap[deleteID];
+            nodeMap.erase(deleteID);
+            for(auto&& i : temp->neighbors){
+                list<Node*>::iterator it = i->neighbors.begin();
+                while(it != i->neighbors.end()){
+                    if ((*it)->nodeID == temp->nodeID){
+                        it = i->neighbors.erase(it);
+                    }
+                    else{
+                        it++;
+                    }
+
+                }
```

Comment:
You would have better complexity and cleaner logic here by using `std::set<Node*>` to store the neighbors. (Since the order of neighbors is irrelevant and this isn't a multigraph data structure)